doc/fileformat/nmodl.rst

```diff
@@ -191,3 +191,120 @@ contain full ``segments``).
+linked into Arbor (the built-in catalogues).
+
+Now, we turn to a series of tips we found helpful in producing fast NMODL
+mechanisms.
+
+``RANGE``
+---------
+
+Parameters and ``ASSIGNED`` variables marked as ``RANGE`` will be stored as an
+array with one entry per CV in Arbor. Reading and writing these incurs a memory
+access and thus affects cache and memory utilisation metrics. It is often
+cheaper to use ``LOCAL`` variables instead, even if that means foregoing the
+ability to re-use a comuted value. Compute is so much faster than memory on
+modern hardware that re-use at the expense of memory accesses is seldom
+profitable, except for the most complex terms.
+
+``PROCEDURE``
+-------------
+
+Prefer ``FUNCTION`` over ``PROCEDURE``. The latter *require* ``ASSIGNED RANGE``
+variables to return values and thus stress the memory system for little gain, as
+noted above. Also, they may not be inlined, as opposed to a ``FUNCTION``.
+
+```PARAMETER``
+--------------
+
+``PARAMETER`` should only be used for values that must be set by the simulator.
+All fixed values should be ``CONSTANT`` instead. These can be inlined and folded
+into the computations, allowing for further optimisations.
+
+Sharing Expressions Between ``INITIAL`` and ``BREAKPOINT`` or ``DERIVATIVE``
+----------------------------------------------------------------------------
+
+This is often done using a ``PROCEDURE``, which we know is inefficient. On top,
+this ``PROCEDURE`` will likely compute more outputs than strictly needed to
+accomodate both blocks. DRY code is a good idea nevertheless, so use a series of
+``FUNCTION`` instead to compute common expressions.
+
+This leads naturally to a common optimisation in H-H style ion channels. If you
+heeded the advice above, you will likely see this patter emerge:
+
+.. code::
+
+   na   = n_alpha()
+   nb   = n_beta()
+	 ntau = 1/(na + nb)
+   ninf = na*ntau
+
+   n' = (ninf - n)/ntau
+
+Written out in this explicit way it becomes obvious that this can be expressed
+in a more compact way
+
+.. code::
+
+   na   = n_alpha()
+   nb   = n_beta()
+	 nrho = na + nb
+
+   n' = (na - n)*nrho
+
+The latter code is much faster and neither ``modcc`` nor the external C++ might
+pick it up as an optimisations, largely due to floating point accuracy
+requirements. This is less easy to see when partially hidden in a ``PROCEDURE``.
+
+The Conductance Trick
+---------------------
+
+``modcc``, Arbor's NMODL compiler, applies symbolic differentiation to the
+current expression to find the conductance as ``g = d U/d I`` which are then
+used to compute the voltage update. ``g`` is thus computed multiple times every
+timestep and if the corresponding expression is inefficient, it will cost more
+time than needed. The differentiation implementation is less than smart and will
+not optimise the resulting code. This is an internal detail of Arbor, but it can
+help to produce well performing code. Here is an example
```

Comment:
```suggestion
help to produce well performing code by extracting and separately storing constant values. Here is an example
```

Reply:
That's redundant.

Reply:
Which is a good thing.

Reply:
Not here, it makes the sentence longer, is not really to the point. The sentence before talks about the symdiff 'engine' and its lack of cleverness. _That_ is the bit of knowledge that result in the following being an optimisation.